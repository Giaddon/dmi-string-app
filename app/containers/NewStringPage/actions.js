import {
  CHANGE_STRING,
  ADD_STRING,
  ADD_STRING_SUCCESS,
  ADD_STRING_FAILURE,
} from './constants';

/**
 * Dispatched when the user makes a change to the new string input.
 *
 * @param {string} newString The submitted string
 *
 * @return {object} An action object with a type of CHANGE_STRING and the new string.
 */

export function changeString(newString) {
  return {
    type: CHANGE_STRING,
    newString,
  };
}

/**
 * Dispatched when the user submits a new string for the server.
 *
 * @return {object} An action object with a type of ADD_STRING.
 */

export function addString() {
  return { type: ADD_STRING };
}
/**
 * Dispatched when the new string is added to ther server successfully.
 *
 * @param {string} newString The newly-added string sent from the server
 *
 * @return {object} An action object with a type of ADD_STRINGS_SUCCESS and the server response.
 */
export function addStringSuccess(newString) {
  return { type: ADD_STRING_SUCCESS, newString };
}
/**
 * Dispatched when there's an error adding a new string to the server.
 *
 * @param {object} error The error
 *
 * @return {object} An action object with a type of ADD_STRING_FAILURE and the error.
 */
export function addStringFailure(error) {
  return { type: ADD_STRING_FAILURE, error };
}
