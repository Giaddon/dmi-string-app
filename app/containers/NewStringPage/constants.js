/** Action types for NewStringPage */

export const CHANGE_STRING = 'dmi/NewStringPage/CHANGE_STRING';
export const ADD_STRING = 'dmi/NewStringPage/ADD_STRING';
export const ADD_STRING_SUCCESS = 'dmi/NewStringPage/ADD_STRING_SUCCESS';
export const ADD_STRING_FAILURE = 'dmi/NewStringPage/ADD_STRING_FAILURE';
