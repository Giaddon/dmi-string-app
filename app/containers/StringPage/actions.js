import {
  GET_STRINGS,
  GET_STRINGS_SUCCESS,
  GET_STRINGS_FAILURE,
} from './constants';

/**
 * Requests string data from the server.
 *
 * @return {object} An action object with a type of GET_STRINGS
 */
export function getStrings() {
  return { type: GET_STRINGS };
}

/**
 * Dispatched when string data recieved from the server.
 *
 * @param  {array} strings An array of strings recieved from the server.
 *
 * @return {object} An action object with a type of GET_STRINGS_SUCCESS and the string array
 */
export function getStringsSuccess(strings) {
  return { type: GET_STRINGS_SUCCESS, strings };
}

/**
 * Dispatched when there's an error retrieving string data from the server.
 *
 * @param {object} error The error
 *
 * @return {object} An action object with a type of GET_STRINGS_FAILURE and the error.
 */
export function getStringsFailure(error) {
  return { type: GET_STRINGS_FAILURE, error };
}
