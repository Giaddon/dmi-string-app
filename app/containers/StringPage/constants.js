/** Action types for StringPage */

export const GET_STRINGS = 'dmi/StringPage/GET_STRINGS';
export const GET_STRINGS_SUCCESS = 'dmi/StringPage/GET_STRINGS_SUCCESS';
export const GET_STRINGS_FAILURE = 'dmi/StringPage/GET_STRINGS_FAILURE';
