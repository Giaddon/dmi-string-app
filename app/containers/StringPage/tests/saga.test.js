/**
 * Tests for StringPage sagas
 */

import { put, takeLatest } from 'redux-saga/effects';

import { GET_STRINGS } from '../constants';
import { getStringsSuccess, getStringsFailure } from '../actions';

import getStringsSaga, { getStringsFromServer } from '../saga';

/* eslint-disable redux-saga/yield-effects */
describe('getStringsFromServer Saga', () => {
  let getStringsGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getStringsGenerator = getStringsFromServer();
    const selectDescriptor = getStringsGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();
  });

  it('should dispatch the getStringsSuccess action if it recieves the data successfully', () => {
    const response = ['TEST STRING1', 'TEST STRING2'];
    const putDescriptor = getStringsGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(getStringsSuccess(response)));
  });

  it('should call the getStringsFailure action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getStringsGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(getStringsFailure(response)));
  });
});

describe('getStrings Saga', () => {
  const getStringsSagaFunc = getStringsSaga();

  it('should start task to watch for GET_STRINGS action', () => {
    const takeLatestDescriptor = getStringsSagaFunc.next().value;
    expect(takeLatestDescriptor).toEqual(
      takeLatest(GET_STRINGS, getStringsFromServer),
    );
  });
});
