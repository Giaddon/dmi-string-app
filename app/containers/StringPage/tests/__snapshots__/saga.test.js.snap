// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getStringsFromServer Saga should call the getStringsFailure action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getStringsFromServer Saga should dispatch the getStringsSuccess action if it recieves the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
